[FX] Make `immutable_list` and `immutable_dict` work with pytrees

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73766

`immutable_list` and `immutable_dict` should work with pytree

Differential Revision: [D34630217](https://our.internmc.facebook.com/intern/diff/D34630217)

Commit:
[FX] Make `immutable_list` and `immutable_dict` work with pytrees

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34630217).

Commit:
Update on "[FX] Make `immutable_list` and `immutable_dict` work with pytrees"


`immutable_list` and `immutable_dict` should work with pytree

Differential Revision: [D34630217](https://our.internmc.facebook.com/intern/diff/D34630217)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34630217).

Chillee reviewed (approved):
LGTM (modulo some minor things we resolved offline).

Commit:
Update on "[FX] Make `immutable_list` and `immutable_dict` work with pytrees"


`immutable_list` and `immutable_dict` should work with pytree

Differential Revision: [D34630217](https://our.internmc.facebook.com/intern/diff/D34630217)

[ghstack-poisoned]

zou3519 commented on torch/fx/immutable_collections.py:
```diff
@@ -26,3 +29,22 @@ def _create_immutable_container(base, mutable_functions):
 compatibility(is_backward_compatible=True)(immutable_dict)
+
+
+# Register immutable collections for PyTree operations
```
These should go in torch/utils/_pytree, right?
- It would be good to have all the definitions near each other (https://github.com/pytorch/pytorch/blob/master/torch/utils/_pytree.py#L65)
- I don't know if this file gets imported when someone uses torch.utils._pytree or if it only gets imported when someone calls `import torch.fx`. We want general users of pytorch's PyTree to be able to benefit from these, not just when they `import torch.fx`.

zou3519 replied:
Oh, nevermind, immutable list/dict aren't python constructs, they're defined here. Please ignore.
